crates/net/src/websocket/futures.rs

```diff
@@ -70,8 +70,49 @@ impl WebSocket {
     /// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket#exceptions_thrown)
     /// to learn more.
     pub fn open(url: &str) -> Result<Self, JsError> {
+        Self::setup(web_sys::WebSocket::new(url))
+    }
+
+    /// Establish a WebSocket connection.
+    ///
+    /// This function may error in the following cases:
+    /// - The port to which the connection is being attempted is being blocked.
+    /// - The URL is invalid.
+    /// - The specified protocol is not supported
+    ///
+    /// The error returned is [`JsError`]. See the
+    /// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket#exceptions_thrown)
+    /// to learn more.
+    pub fn open_with_protocol(url: &str, protocol: &str) -> Result<Self, JsError> {
+        Self::setup(web_sys::WebSocket::new_with_str(url, protocol))
+    }
+
+    /// Establish a WebSocket connection.
+    ///
+    /// This function may error in the following cases:
+    /// - The port to which the connection is being attempted is being blocked.
+    /// - The URL is invalid.
+    /// - The specified protocols are not supported
+    /// - The protocols cannot be converted to a JSON string list
+    ///
+    /// The error returned is [`JsError`]. See the
+    /// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket#exceptions_thrown)
+    /// to learn more.
+    pub fn open_with_protocols<S: AsRef<str> + serde::Serialize>(
```

Comment:
Why is `AsRef<str>` neededd

Reply:
I wanted to allow any of:

-  `&[&str]`
- `&[String]`
- `&[arcstr::ArcStr]`

Just allowing a slice of any type which implements `serde::Serialize` means that the user can pass a slice of integers, which will cause an error to be raised, whereas constraining it catches more errors at compile time.

I considered constraining it to only support `&[&str]`, but if the protocol names are generated at runtime, e.g. a `Vec<String>`, then this would involve creating a `Vec<&str>` as well which seems unnecessary. I thus wanted "a slice of things which serialize as strings", and `AsRef<str>` seemed the best match of "thing which serializes as a string"